### PR TITLE
docs: added Algolia search to website

### LINKED ISSUE #4165: Add Algolia search to the website
Splitting out of #4156 as this has been in the works for a while: we need a good docs search solution on the site. Algolia is high quality, free for our kind of open source site, and the de facto standard for Docusaurus.

### PR DESCRIPTION
**Pull request by JoshuaKGoldberg**
## PR Checklist

-   [x] Addresses an existing issue: fixes #4165
-   [x] That issue was marked as [accepting prs](https://github.com/typescript-eslint/typescript-eslint/issues?q=is%3Aopen+is%3Aissue+label%3A%22accepting+prs%22)
-   [x] Steps in [CONTRIBUTING.md](https://github.com/typescript-eslint/typescript-eslint/blob/master/CONTRIBUTING.md) were taken

## Overview

Follows https://docusaurus.io/docs/search#using-algolia-docsearch to get Algolia search on the site. 🔍 

~In my usage this doesn't change the website right now -no header search bar or `/search` page-, so keeping in draft... I don't know why it's not showing locally. The steps in https://github.com/facebook/docusaurus/issues/4975, run in both the repo root and the `packages/website` directory aren't fixing it, and I don't see any obvious typos or missing packages.~ ✔️

### CODE REVIEW

File: packages/website/docusaurus.config.js [hunk 1]
```diff
@@ -57,7 +57,11 @@ const config = {
   themeConfig:
     /** @type {import('@docusaurus/theme-common').UserThemeConfig} */
     ({
-      // sidebarCollapsible: false,
+      algolia: {
+        appId: 'BH4D9OD16A',
+        apiKey: '1ad6b47d4e742c4c0653877b5511c602',
+        indexName: 'typescript-eslint',
+      },
```
Comment (JoshuaKGoldberg) [inline packages/website/docusaurus.config.js:64]: @Josh-Cena is there something obviously wrong that I'm missing here...? No search is showing up 😞

Comment (Josh-Cena) [reply packages/website/docusaurus.config.js:64]: Well yeah, you didn't [add `theme-algolia` to the config](https://docusaurus.io/docs/next/using-themes#using-themes). If you are use `preset-classic` this should be done automatically, but I'm a bit surprised (also asked in #4157) that you didn't actually use the preset

Comment (JoshuaKGoldberg) [reply packages/website/docusaurus.config.js:64]: Aha thank you, now it all clicks! That fixed it.

Yeah I don't have a personal preference for or against preset-classic -- but it makes sense why we could use it to standardize+simplify...

File: packages/website/docusaurus.config.js [hunk 2]
```diff
@@ -32,6 +32,7 @@ const config = {
         customCss: require.resolve('./src/css/custom.css'),
       },
     ],
+    '@docusaurus/theme-search-algolia',
```
Comment (Josh-Cena) [inline packages/website/docusaurus.config.js:35]: Just a kind notice: we may introduce shorthands in the next (few) release (facebook/docusaurus#5930) like ESLint's `eslint-plugin-react` -> `plugins: ['react']`. In that case you may want to put this in `themes` instead so you can shorten it to `themes: ['search-algolia']`

**Review by JamesHenry — CHANGES_REQUESTED** [review]
Sorry if I'm missing something but I don't see any search functionality on the latest preview build on this PR?

File: packages/website/src/pages/index.tsx [hunk 3]
```diff
@@ -16,7 +16,8 @@ interface FeatureItem {
 
 const features: FeatureItem[] = [
   {
-    title: 'What are ESLint and TypeScript, and how do they compare?',
+    title:
+      'Hello world: What are ESLint and TypeScript, and how do they compare?',
```
Comment (JoshuaKGoldberg) [inline packages/website/src/pages/index.tsx]: @JamesHenry very odd, I saw the search on my local deployment from the commit in the PR:

Screenshot of local deployment with search

Now, with this `Hello World:` change, I see it on the Netlify deployment too:

Screenshot of Netlify deployment with Hello World and search

🤷

Comment (Josh-Cena) [reply packages/website/src/pages/index.tsx]: Might be related to caching... I can't pinpoint it either

**Comment by bradzacher** [comment]
Just checked the preview deployment and it's all working for me!
Was able to search up a rule and navigate to it.

Great work!

**Review by bradzacher — APPROVED** [review]

**Pull request merged**